src/devices/src/legacy/rtc_pl031.rs

```diff
@@ -92,6 +91,33 @@ impl RTC {
         (ts / utils::time::NANOS_PER_SECOND as i128) as u32
     }
 
+    fn handle_read(&mut self, offset: u64) -> Result<u32> {
+        let val;
+
+        if offset < AMBA_ID_HIGH && offset >= AMBA_ID_LOW {
+            let index = ((offset - AMBA_ID_LOW) >> 2) as usize;
+            val = u32::from(PL031_ID[index]);
+        } else {
+            val = match offset {
+                RTCDR => self.get_time(),
+                RTCMR => {
+                    METRICS.rtc.missed_read_count.inc();
+                    // Even though we are not implementing RTC alarm we return the last value
+                    self.match_value
+                }
+                RTCLR => self.load,
+                RTCCR => 1, // RTC is always enabled.
+                RTCIMSC => self.imsc,
+                RTCRIS => self.ris,
+                RTCMIS => self.ris & self.imsc,
+                off => {
+                    return Err(Error::BadOffset(off));
+                }
+            };
+        }
+        Ok(val)
+    }
+
     fn handle_write(&mut self, offset: u64, val: u32) -> Result<()> {
         match offset {
```

Comment:
Not really tied to this PR, but the spec says that writing 1 to the RTCICR register clears the interrupt.
Writing 0 should not.
https://developer.arm.com/documentation/ddi0224/c/Programmers-model/General-registers/Interrupt-Clear-Register--RTCICR?lang=en

Reply:
Also, register offsets are declared as u64, when I think they can even be of type `u8`

Reply:
> Not really tied to this PR, but the spec says that writing 1 to the RTCICR register clears the interrupt.
> Writing 0 should not.
> https://developer.arm.com/documentation/ddi0224/c/Programmers-model/General-registers/Interrupt-Clear-Register--RTCICR?lang=en

> Not really tied to this PR, but the spec says that writing 1 to the RTCICR register clears the interrupt.
> Writing 0 should not.
> https://developer.arm.com/documentation/ddi0224/c/Programmers-model/General-registers/Interrupt-Clear-Register--RTCICR?lang=en

There is a problem with the RTCICR, but it is not related to how we set it. The RTCICR is used in conjunction with the RTCMR to implement a  time alarm function:
 ```
The Match Register is programmed by writing to RTCMR. The counter and match values are compared in a comparator. When both values are equal, the RTCINTR interrupt is asserted HIGH. An external processor can use the interrupt to implement a basic time alarm function. The interrupt is cleared by writing any data value to the Interrupt Clear Register, RTCICR. The value in the Match Register can be read at any time.
```
This is extra functionality that one can achieve with the RTC. As a consequence, our implementation does not need the RTCICR. It can easily be seen from the implementation that the value of self.ris never changes, so there is no need for the value to be cleared by the RTCICR as it will always be 0. I updated the code to be similar to how we proceed for the RTCMR register, take a look.

Reply:
> Also, register offsets are declared as u64, when I think they can even be of type `u8`

Indeed the register offsets could be u8, but our interface works with u64. The offsets are u64 cause the `MmioRead(u64 /* address */, &'a mut [u8]),` reads it in an u64 and gives the value further to the `read` and `write` bus functions.

Reply:
> > Not really tied to this PR, but the spec says that writing 1 to the RTCICR register clears the interrupt.
> > Writing 0 should not.
> > https://developer.arm.com/documentation/ddi0224/c/Programmers-model/General-registers/Interrupt-Clear-Register--RTCICR?lang=en
> 
> > Not really tied to this PR, but the spec says that writing 1 to the RTCICR register clears the interrupt.
> > Writing 0 should not.
> > https://developer.arm.com/documentation/ddi0224/c/Programmers-model/General-registers/Interrupt-Clear-Register--RTCICR?lang=en
> 
> There is a problem with the RTCICR, but it is not related to how we set it. The RTCICR is used in conjunction with the RTCMR to implement a time alarm function:
> 
> ```
> The Match Register is programmed by writing to RTCMR. The counter and match values are compared in a comparator. When both values are equal, the RTCINTR interrupt is asserted HIGH. An external processor can use the interrupt to implement a basic time alarm function. The interrupt is cleared by writing any data value to the Interrupt Clear Register, RTCICR. The value in the Match Register can be read at any time.
> ```
> 
> This is extra functionality that one can achieve with the RTC. As a consequence, our implementation does not need the RTCICR. It can easily be seen from the implementation that the value of self.ris never changes, so there is no need for the value to be cleared by the RTCICR as it will always be 0. I updated the code to be similar to how we proceed for the RTCMR register, take a look.


In the latest commit, I see that this was fixed.

Since we are not using the match functionality, can't we get rid of all interrupts?

Reply:
> > Also, register offsets are declared as u64, when I think they can even be of type `u8`
> 
> Indeed the register offsets could be u8, but our interface works with u64. The offsets are u64 cause the `MmioRead(u64 /* address */, &'a mut [u8]),` reads it in an u64 and gives the value further to the `read` and `write` bus functions.

I see. I think a cast should do it though, right?

Reply:
> > > Not really tied to this PR, but the spec says that writing 1 to the RTCICR register clears the interrupt.
> > > Writing 0 should not.
> > > https://developer.arm.com/documentation/ddi0224/c/Programmers-model/General-registers/Interrupt-Clear-Register--RTCICR?lang=en
> > 
> > 
> > > Not really tied to this PR, but the spec says that writing 1 to the RTCICR register clears the interrupt.
> > > Writing 0 should not.
> > > https://developer.arm.com/documentation/ddi0224/c/Programmers-model/General-registers/Interrupt-Clear-Register--RTCICR?lang=en
> > 
> > 
> > There is a problem with the RTCICR, but it is not related to how we set it. The RTCICR is used in conjunction with the RTCMR to implement a time alarm function:
> > ```
> > The Match Register is programmed by writing to RTCMR. The counter and match values are compared in a comparator. When both values are equal, the RTCINTR interrupt is asserted HIGH. An external processor can use the interrupt to implement a basic time alarm function. The interrupt is cleared by writing any data value to the Interrupt Clear Register, RTCICR. The value in the Match Register can be read at any time.
> > ```
> > 
> > 
> > This is extra functionality that one can achieve with the RTC. As a consequence, our implementation does not need the RTCICR. It can easily be seen from the implementation that the value of self.ris never changes, so there is no need for the value to be cleared by the RTCICR as it will always be 0. I updated the code to be similar to how we proceed for the RTCMR register, take a look.
> 
> In the latest commit, I see that this was fixed.
> 
> Since we are not using the match functionality, can't we get rid of all interrupts?

Yup, I got rid of the interrupts. PTAL!

Reply:
Regarding the offsets, u8 does not seem to be enough, the last address from the RTC spec is `0xFFC RTCPCellID3` which cannot be represented in an u8. Of course, we could use an u16, let me know if you have a strong preference for doing a cast to the smallest possible representation.

Reply:
I don't have a very strong preference. I think it probably only adds a tiny memory overhead